modal/_functions.py

```diff
@@ -256,45 +258,50 @@ async def _retry_input(self) -> None:
                 clear_on_success=True,
                 input_jwts=[expected_jwt] if expected_jwt else None,
             )
         ).outputs[0]
-        return await _process_result(item.result, item.data_format, self.stub, self.client)
+        return await _process_result_wrap_exceptions(item.result, item.data_format, self.stub, self.client)
 
     async def run_function(self) -> Any:
-        # Use retry logic only if retry policy is specified and
-        ctx = self._retry_context
-        if (
-            not ctx
-            or not ctx.retry_policy
-            or ctx.retry_policy.retries == 0
-            or ctx.function_call_invocation_type != api_pb2.FUNCTION_CALL_INVOCATION_TYPE_SYNC
-            or not ctx.sync_client_retries_enabled
-        ):
-            return await self._get_single_output()
+        try:
+            # Use retry logic only if retry policy is specified and
+            ctx = self._retry_context
+            if (
+                not ctx
+                or not ctx.retry_policy
+                or ctx.retry_policy.retries == 0
+                or ctx.function_call_invocation_type != api_pb2.FUNCTION_CALL_INVOCATION_TYPE_SYNC
+                or not ctx.sync_client_retries_enabled
+            ):
+                return await self._get_single_output_wrap_exceptions()
 
-        # User errors including timeouts are managed by the user specified retry policy.
-        user_retry_manager = RetryManager(ctx.retry_policy)
+            # User errors including timeouts are managed by the user specified retry policy.
+            user_retry_manager = RetryManager(ctx.retry_policy)
 
-        while True:
-            try:
-                return await self._get_single_output(ctx.input_jwt)
-            except (UserCodeException, FunctionTimeoutError) as exc:
-                delay_ms = user_retry_manager.get_delay_ms()
-                if delay_ms is None:
-                    raise exc
-                await asyncio.sleep(delay_ms / 1000)
-            except InternalFailure:
-                # For system failures on the server, we retry immediately,
-                # and the failure does not count towards the retry policy.
-                pass
-            await self._retry_input()
+            while True:
+                try:
+                    return await self._get_single_output_wrap_exceptions(ctx.input_jwt)
+                except (_ContainerException, FunctionTimeoutError):
+                    delay_ms = user_retry_manager.get_delay_ms()
+                    if delay_ms is None:
+                        raise
+                    await asyncio.sleep(delay_ms / 1000)
+                except InternalFailure:
+                    # For system failures on the server, we retry immediately,
+                    # and the failure does not count towards the retry policy.
+                    pass
+
+                await self._retry_input()
+        except _ContainerException as exc:
+            raise exc.unwrap()
```

Comment:
Nit: In an effort to try to reduce the amount of indenting can the `try`, `except` be a decorator? Something like this:

```python
def _unwrap_container_exception(func):
    @functools.wraps(func)
    async def wrapper(*args, **kwargs):
        try:
            return await func(*args, **kwargs)
        except _ContainerException as exc:
            raise exc.unwrap()
    return wrapper

@_unwrap_container_exception
async def run_function(self) -> Any:
    ...
```